src/python_routing_v02/diffusive_utils.py

```diff
@@ -0,0 +1,655 @@
+# Assumption: top segments of headbasin reaches get their own lateral flows as 
+#             upstream boundary conditions while making these qlateral zeroes
+#------------------------------------------------------------------------------------------
+def fp_ubcd_map(frnw_g
+            , pynw
+            , nts_ub_g
+            , nrch_g
+            , geo_index
+            , qlat_data
+            , qlat_g
+            ):   
+    
+    ubcd_g=np.zeros((nts_ub_g, nrch_g)) 
+    frj= -1
+    for frj in range(nrch_g):
+        if frnw_g[frj,2]==0: # the number of upstream reaches is zero.
+            head_segment=pynw[frj]
+            for tsi in range(0,nts_ub_g):
+                
+                idx_segID = np.where(geo_index == head_segment)
+                
+                ubcd_g[tsi, frj]= qlat_data[idx_segID,tsi] # [m^3/s]
+                qlat_g[tsi,0,frj]=0.0                   
+    
+    return ubcd_g 
+                        
+#------------------------------------------------------------------------------------------
+#       downstream boundary condition mapping between Python and Fortran
+# 
+#   ** In this mapping, USGS observed stage data at TW is used.
+#------------------------------------------------------------------------------------------
+def fp_dbcd_map(usgsID2tw
+                , usgssDT
+                , usgseDT
+                , usgspCd
+                ):
+   
+    
+    # ** 1) downstream stage (here, lake elevation) boundary condition
+    #from nwis_client.iv import IVDataService
+    from hydrotools.nwis_client.iv import IVDataService
+    #from evaluation_tools.nwis_client.iv import IVDataService
+    # Retrieve streamflow and stage data from two sites
+        # Note: 1. Retrieved data all are based on UTC time zone (UTC is 4 hours ahead of Eastern Time during
+        #          daylight saving time and 5 hours ahead during standard time) 
+        #       2. Retrieved data are always 1 hour ahead of stated startDT and 1 hour ahead of endDT,
+        #          where starDT or endDT equal to yyyy-mm-dd 00:00.
+        #       3. Also, retrieved data in 15 min so there are always four more data before startDT
+        #          and four less data before endDT.
+        #          For example, startDT='2018-08-01' and endDT='2018-09-01', then the retrieved data starts by
+        #          2018-07-31-23:00:00
+        #          2018-07-31-23:15:00
+        #          2018-07-31-23:30:00
+        #          2018-07-31-23:45:00
+        #          2018-08-01-00:00:00
+        #               .......
+        #          2018-08-31-22:00:00
+        #          2018-08-31-22:15:00
+        #          2018-08-31-22:30:00
+        #          2018-08-31-22:45:00
+        #          2018-08-31-23:00:00        
+        #       4. '00060' for discharge [ft^3/s]  
+        #          '00065' for stage [ft]
+        #          '62614' for Elevation, lake/res,NGVD29 [ft]
+    observations_data = IVDataService.get(
```

Comment:
probably need to wrap this with some error handling, in the case NWIS is down or some other issue exists in getting the data, what should happen?

Reply:
I would shy away from sticking the error handling in here -- it would only be a stopgap to hold us over until there is a broader function for obtaining the downstream boundary timeseries which would take options to seek USGS data or other time series from files in the directory. All that would be presented to the diffusive function all cleaned up so that the data getting isn't in this utility file at all...

Reply:
We will aim to eventually pull the DL of USGS data from this script and into something stand alone. Putting this issue on the shelf for a little while. 

Will want to understand the operational context of diffusive wave before we start making these decisions.